fault.h

```diff
@@ -89,6 +128,127 @@ struct kibosh_fault_read_delay {
     double fraction;
 };
 
+/**
+ * The class for Kibosh faults that make files unwritable.
+ */
+struct kibosh_fault_unwritable {
+    /**
+     * The base class members.
+     */
+    struct kibosh_fault_base base;
+
+    /**
+     * The path prefix, starts with '/'.
+     */
+    char *prefix;
+
+    /**
+     * The type of file to be made unwritable.
+     */
+    char *file_type;
+
+    /**
+     * The error code to return from write faults.
+     */
+    int code;
+};
+
+/**
+ * The class for Kibosh faults that lead to corrupted data when reading.
+ */
+struct kibosh_fault_read_corrupt {
+    /**
+     * The base class members.
+     */
+    struct kibosh_fault_base base;
+
+    /**
+     * The path prefix, starts with '/'.
+     */
+    char *prefix;
+
+    /**
+     * The type of file to be corrupted.
+     */
+    char *file_type;
+
+    /**
+     * Mode of byte corruption.
+     * 1000 -> zeros (default)
+     * 1001 -> random values
+     * 1100 -> sequential zeros
+     * 1101 -> sequential random values
+     * 1200 -> silently drop a fraction of bytes at the end of buffer
+     */
+    int mode;
+
+    /**
+     * Number of corruption fault injected before switching to unwritable fault.
+     * Less than 0 means never switch.
+     */
+    int count;
+
+    /**
+     * The fraction of bytes to be corrupted.
+     */
+    double fraction;
+
+    /**
+     * Whether to store corrupted data to log.
+     * 0 means do not store data.
+     * Greater than 0 means store data to log.
+     */
+    int store_data;
+};
+
+/**
+ * The class for Kibosh faults that lead to corrupted data when writing.
+ */
+struct kibosh_fault_write_corrupt {
+    /**
+     * The base class members.
+     */
+    struct kibosh_fault_base base;
+
+    /**
+     * The path prefix, starts with '/'.
+     */
+    char *prefix;
+
+    /**
+    * The type of file to be corrupted.
+    */
+    char *file_type;
+
+    /**
+     * Mode of byte corruption.
+     * 1000 -> zeros (default)
+     * 1001 -> random values
+     * 1100 -> sequential zeros
```

Comment:
Could we elaborate a bit on the difference between the `sequential` and non-sequential variants?

Reply:
Sequential corruption will only corrupt the continuous portion of the file near the end, depending on the fraction set on the fault. Non-sequential variants will corrupt random positions in a file. So sequential corruption was intended to simulate a power fail scenario where the end of a file might not be written to disk correctly. Non-sequential corruption corrupts random bytes of the file and can mimic other disk corruptions.

Reply:
I understand that. I meant that we should add a comment in the code about the difference.

Reply:
@yyu1993 could you add this comment to the code?

Reply:
Comment was added.